src/app/views/query-runner/query-input/AutoComplete.tsx

```diff
@@ -0,0 +1,264 @@
+      suggestions: [],
+      showSuggestions: false,
+      userInput: this.props.sampleQuery.sampleUrl,
+      compare: ''
+    };
+  }
+
+  public onChange = (e: any) => {
+    const { suggestions, showSuggestions, userInput: previousUserInput, compare } = this.state;
+    const userInput = e.target.value;
+
+    this.props.contentChanged(userInput);
+
+    this.setState({
+      userInput
+    });
+
+    if (showSuggestions && suggestions.length) {
+      this.filterSuggestions(userInput, previousUserInput, compare, suggestions);
+    }
+    this.initialiseAutoComplete(userInput);
+  };
+
+  public onClick = (e: any) => {
+    const selected = e.currentTarget.innerText;
+    this.appendSuggestionToUrl(selected);
+  };
+
+  private initialiseAutoComplete = (url: string) => {
+    const lastCharacter = url.substring(url.length - 1);
+    switch (lastCharacter) {
+      case '?':
+        this.requestForAutocompleteOptions(url);
+        break;
+
+      case '=':
+        this.getParameterEnums(url);
+        break;
+
+      case ',':
+        this.getParameterEnums(url);
+        break;
+
+      case '&':
+        this.getPathParameters();
+        break;
+
+      default:
+        break;
+    }
+  }
+
+  public onKeyDown = (e: any) => {
+    const { activeSuggestion, filteredSuggestions, showSuggestions } = this.state;
+
+    switch (e.keyCode) {
+      case KeyCodes.tab:
+        if (showSuggestions) {
+          const selected = filteredSuggestions[activeSuggestion];
+          this.appendSuggestionToUrl(selected);
+        }
+        break;
+
+      case KeyCodes.up:
+        if (activeSuggestion === 0) {
+          return;
+        }
+        this.setState({ activeSuggestion: activeSuggestion - 1 });
+        break;
+
+      case KeyCodes.down:
+        if (activeSuggestion - 1 === filteredSuggestions.length) {
+          return;
+        }
+        this.setState({ activeSuggestion: activeSuggestion + 1 });
+        break;
+
+      default:
+        break;
+    }
+
+  };
+
+  public getPathParameters = () => {
+    const { parameters } = this.props.autoCompleteOptions;
+    const suggestions: string[] = [];
+    if (parameters) {
+      parameters.forEach((element: any) => {
+        suggestions.push(element.name);
+      });
+    }
+
+    this.setState({
+      filteredSuggestions: suggestions,
+      suggestions,
+      showSuggestions: true,
+      compare: ''
+    });
+  }
+
+  public componentDidUpdate = (prevProps: IAutoCompleteProps) => {
+    if (prevProps.autoCompleteOptions !== this.props.autoCompleteOptions) {
+      this.getPathParameters();
+    }
+  }
+
+  private getParameterEnums = (url: string) => {
+    const param = url.split('$').pop()!.split('=')[0];
+    const { parameters } = this.props.autoCompleteOptions;
+    const section = parameters.find(k => k.name === `$${param}`);
+    const list: string[] = [];
+    if (section.items.length > 0) {
+      section.items.forEach((element: string) => {
```

Comment:
I don't think this loop is necessary. What if you just had:
```
let list: string[] = section.items;
```
I think it would boost its runtime, what do you think?